unattended-upgrade

```diff
@@ -1265,86 +1305,97 @@ def get_dpkg_log_content(logfile_dpkg, install_start_time):
 
 def get_auto_removable(cache):
     # type: (apt.Cache) -> AbstractSet[str]
-    return set([pkg.name for pkg in cache
-                if pkg.is_auto_removable])
-
-
-def only_kernel_pkgs(auto_removable):
-    # type: (AbstractSet[str]) -> AbstractSet[str]
-    running_kernel_version = subprocess.check_output(
-        ["uname", "-r"], universal_newlines=True).rstrip()
-    versioned_kernel_pkgs = apt_pkg.config.value_list(
-        "APT::VersionedKernelPackages")
-
-    if versioned_kernel_pkgs:
-        kernel_pkgs_regexp = "(" + "|".join(
-            ["^" + p for p in versioned_kernel_pkgs]) + ")"
-        logging.debug("Using %s regexp to find kernel packages",
-                      kernel_pkgs_regexp)
-
-        return set([pkg for pkg in auto_removable
-                    if (re.match(kernel_pkgs_regexp, pkg) and
-                        not re.match(running_kernel_version, pkg))])
-    else:
-        logging.debug("APT::VersionedKernelPackages is not set")
-        return set()
+    return {pkg.name for pkg in cache
+            if pkg.is_auto_removable}
 
 
-def is_autoremove_valid(cache, pkgname, auto_removable):
-    # type: (apt.Cache, str, AbstractSet[str]) -> bool
-    changes = set([p.name for p in cache.get_changes()])
+def is_autoremove_valid(cache, pkgname, auto_removable, blacklisted_pkgs,
+                        whitelisted_pkgs):
+    # type: (apt.Cache, str, AbstractSet[str], List[str], List[str]) -> bool
+    changes = cache.get_changes()
     if not changes:
         # package is already removed
         return True
-    if not changes.issubset(auto_removable):
+    pkgnames = {pkg.name for pkg in changes}
+    for pkg in changes:
+        if not is_pkg_change_allowed(pkg, blacklisted_pkgs, whitelisted_pkgs):
+            logging.warning(
+                _("Keeping the following auto-removable package(s) because "
+                  "they include %s which is set to be kept unmodified: %s"),
+                pkg.name, " ".join(sorted(pkgnames)))
+            return False
+    if not pkgnames.issubset(auto_removable):
         if pkgname != "":
             logging.warning(
                 _("Keeping auto-removable %s package(s) because it would"
                   " also remove the following packages which should "
                   "be kept in this step: %s"), pkgname,
-                " ".join(sorted(changes - auto_removable)))
+                " ".join(sorted(pkgnames - auto_removable)))
         else:
             logging.warning(
                 _("Keeping %s auto-removable package(s) because it would"
                   " also remove the following packages which should "
                   "be kept in this step: %s"), len(auto_removable),
-                " ".join(sorted(changes - auto_removable)))
+                " ".join(sorted(pkgnames - auto_removable)))
 
         cache.clear()
         return False
+    for packagename in pkgnames:
+        if cache.running_kernel_pkgs_regexp and \
```

Comment:
(nitpick^2): I personally prefer using () instead of \, so: 
```
if (cache.running_kernel... and 
        cache....):
```

Reply:
Me too, I converted some lines to use ... \ to avoid indenting the same amount as the next lines